sdks/python/apache_beam/runners/dataflow/dataflow_runner.py

```diff
@@ -441,22 +443,25 @@ def _get_side_input_encoding(self, input_encoding):
 
   def _get_encoded_output_coder(self, transform_node, window_value=True):
     """Returns the cloud encoding of the coder for the output of a transform."""
+    from apache_beam.runners.dataflow.internal import apiclient
     if (len(transform_node.outputs) == 1
         and transform_node.outputs[None].element_type is not None):
       # TODO(robertwb): Handle type hints for multi-output transforms.
       element_type = transform_node.outputs[None].element_type
+      use_fnapi = apiclient._use_fnapi(transform_node.outputs[None].pipeline._options)
     else:
       # TODO(silviuc): Remove this branch (and assert) when typehints are
       # propagated everywhere. Returning an 'Any' as type hint will trigger
       # usage of the fallback coder (i.e., cPickler).
       element_type = typehints.Any
+      use_fnapi = False  # TODO(chambers): XXX do the right thing for this
```

Comment:
This is unfortunate.  Is there something better I can/should do here?

In general, passing around the use_fnapi flag is yucky.  I'd much rather have the pipeline or the pipeline options be available in an instance variable.  Is there a way to do that?  Or a reason not to do that?

Reply:
This here is just a placeholder.  I don't know how to get access to the pipeline options otherwise.  If you tell me how, I'll fix it.

Reply:
I'd like to understand why setting the `pipeline_proto_coder_id` attribute unconditionally breaks things. If that's not workable, I'd rather name this something other than use_fnapi if we don't need the coder id in this case.

Reply:
One post-commit test failed, on a checksum comparison.  I don't have any deeper understanding of what the test is doing or why there was a failure.  I have had other experiences where tests were (brittlely) checking for equivalence against some expected representation which can be adversely affected by adding an otherwise unused property to CloudObjects.

To be clear, we *do* need the coder id in this case, at least when we support multi-output DoFns over the fnapi using the worker code that reads this property.  We're not running such tests now.  I need advice on how to get a hold of the pipeline object in this branch in order to put in the proper code.  Also, the TODO in this branch suggests that the branch may be going away, so maybe it doesn't need to be fixed.

Reply:
Ping?  What should I do to make progress on this PR?  I'm OK submitting something that (a) doesn't break anything that already exists, and (b) makes some incremental progress on runners using Beam portability.  I'm also happy to improve this CL, if given guidance on what to do.

Reply:
I'd like to understand why this particular test failed, and not others, which may be indicative of other problems, rather than adding a bunch of code to work around it, but at this point we probably shouldn't be blocking on that. 

Let's just rename this something like emit_coder_ids and get it in. 

(I also hope this code in this whole file live on much longer.)

Reply:
I read this as "if there are multiple outputs, don't use the fn api" but I understand your intent now. 

Just thinking about this again, another option would be to simply choose any output on which to check for the fn api flag. This shouldn't(?) be called if there are no outputs. 

As for the test failure, it feels like we're working around a still-present bug. But as you say, it may just be brittle testing. 

As for how this code becomes obsolete, the Dataflow service should just accept FnApi protos directly, rather than have each SDK translate it to cloud objects just to try to get them translated to the right DFE objects on the other end.

Reply:
I've updated this code to just pick the first output to get the pipeline options from.  PTAL.